### ISSUE TITLE
Invalid cross-device link when remove a file in merged directory

### ISSUE THREAD

**Issue by vanloswang** [containers/fuse-overlayfs#33]
My host os is CentOS 7, and I backport fuse3 to it, and compile fuse-overlayfs in the master branch with commit 91bb4019eb448288102c871d820f68843eab3685. The error is occured by following commands:
```
# ls lower
a  aaa  aaa1  aaa2
# ls upper
b  bbb  bbb1  bbb2
# fuse-overlayfs -o lowerdir=lower,upperdir=upper,workdir=work merged
# ls merged/
a  aaa  aaa1  aaa2  b  bbb  bbb1  bbb2
# rm -rf merged/aaa
rm: cannot remove ‘merged/aaa’: Invalid cross-device link
# rm -rf merged/bbb
rm: cannot remove ‘merged/bbb’: Invalid cross-device link
```
lower and upper is encrypt by gocryptfs or encfs, gocryptfs and encfs are fuse filesystem which use fuse high level apis. Take a look at https://github.com/rfjakob/gocryptfs/issues/290 and https://github.com/vgough/encfs/issues/542.
I have test it out that files in lower directory can not be modified, and files in upper directory can be modified.
Fedora 29 is the same with CentOS 7.

**Comment by vanloswang** [comment]
After debug this issue, I get some error info as following:
```
# fuse-overlayfs -f --debug -o lowerdir=lower,upperdir=upper,workdir=work merged
```
```
unique: 151, opcode: LOOKUP (1), nodeid: 1, insize: 52, pid: 10790
   unique: 151, error: -2 (No such file or directory), outsize: 16
unique: 152, opcode: UNLINK (10), nodeid: 1, insize: 44, pid: 10797
   unique: 152, error: -18 (Invalid cross-device link), outsize: 16
```

**Comment by giuseppe** [comment]
I cannot reproduce this issue on Fedora.  Have you changed anything in fuse3 to backport it?

**Comment by vanloswang** [comment]
@giuseppe please follow the commands in https://github.com/VanlOSTech/LayeredCryptFS/blob/master/layeredcryptfs to reproduce this issue.

**Comment by giuseppe** [comment]
@vanloswang do I understand correctly that there is a file system mounted on `upper`?

**Comment by vanloswang** [comment]
@vanloswang yes, it is. A encrypted file system is mounted on ```upper```, it is the same as ```lower``` too. It seems like the encrypted file system does not provide some xattr handles.

**Comment by giuseppe** [comment]
`upper` and `workdir` must be on the same file system and it seems they are not.  It is the same requirement that the kernel implementation of overlay has.  It is needed to "prepare" a file on the `workdir` and then move it atomically to the `upper` directory.

**Comment by dcowan-london** [comment]
Getting the same error when attempting to delete files in a merged dir on Ubuntu 20.04. Upperdir is an sshfs mount, remote server is TrueNAS (so FreeBSD 12.2, filesystem ZFS 2.0.7-1).

Logs when I run `rm`:
```unique: 4, opcode: GETATTR (3), nodeid: 1, insize: 56, pid: 1191
ovl_getattr(ino=1)
   unique: 4, success, outsize: 120
unique: 6, opcode: LOOKUP (1), nodeid: 1, insize: 46, pid: 1191
ovl_lookup(parent=1, name=hello)
   unique: 6, success, outsize: 144
unique: 8, opcode: UNLINK (10), nodeid: 1, insize: 46, pid: 1191
ovl_unlink(parent=1, name=hello)
   unique: 8, error: -18 (Invalid cross-device link), outsize: 16```
